### ISSUE TITLE
Get spiff from ITM$SPIFF_CLNDR

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#218]
_Imported from trac ticket 218.
Created by **kbower**
Opened in trac: Jun 24, 2010, 10:59:10 AM
Last modified in trac: Jan 25, 2011, 9:16:27 AM_
***
Kanes uses the `ITM$SPIFF_CLNDR` table.
```
SQL> desc ITM$SPIFF_CLNDR;
 Name                                Null?    Type
 ----------------------------------- -------- -------------------------
 ITM_CD                              NOT NULL VARCHAR2(9)
 EFF_DT                              NOT NULL DATE
 SPIFF_CD                            NOT NULL VARCHAR2(3)
 END_DT                              NOT NULL DATE
 SPIFF_AMT                           NOT NULL NUMBER(13,2)
 MIN_RETAIL                                   NUMBER(13,2)
 STORE_CD                            NOT NULL VARCHAR2(2)
```
**Implementation Note:** make a joinedload (eagerload) of this relation inside the `Order.attach_all_products()` method.
----
## email:
Hi Wayne,

When an item is on an SO_LN, the SPIFF is copied from the ITM record.  Is this always or are there exceptions?

Also, on each save of the order, (updates) should it do the same or should it only copy the value on SO_LN insert?

Thanks again.
----
Kent,

Good morning, the SPIFF value from the ITM record is copied only on insert into SO_LN. The only exception would be if the SPIFF calendar is being used from ITM$SPIFF_CLNDR where there is effective date, ending date, store code and possible minimum retail.

Thanks,
Wayne Liburd
----
## follow-up
Kent,

If an order is being keyed it should be on SO.SO_WR_DT because there are times when an order is taken before the store is closed or the T1 lines is down and it is keyed the following day but the date of the order is entered as yesterday’s date making the SO.SO_WR_DT yesterday’s date. However, if a line is added to an existing order it should be SYSDATE which will also be the SO_LN.ADDON_WR_DT of the line on that order.

 

Thanks,
Wayne Liburd

From: Kent Bower [mailto:kb@retailarchitects.com]
Sent: Thursday, June 24, 2010 12:49 PM
To: Wayne Liburd
Subject: Re: spiff question

 

Thanks again... just a double check on the SO_WR_DT... this means if someone adds a line a week later to an order, the spiff is still determined from the date the order was written, not based on today?  One could argue it should be based on today's date, so I wanted to double check.

- Kent

On 6/24/2010 11:24 AM, Wayne Liburd wrote:

Kent,

The sorter code would be SO.SO_STORE_CD.
The date would be SO_WR_DT.
MIN_RETAIL would be minimum UNIT_PRC. It should double check the spiff if it is discounted.

Yes, if none is found it defaults to ITM.SPIFF.
Thanks,
Wayne Liburd

From: Kent Bower [mailto:kb@retailarchitects.com]
Sent: Thursday, June 24, 2010 11:10 AM
To: Wayne Liburd
Subject: Re: spiff question

Few follow-up questions:

The Store_cd is which?  (My guess is the SO_STORE_CD on SO, but that might not be the store always (like if customer service is in if adding the line), so I wanted to check.)

The date is which? (The SO_WR_DT or SYSDATE or something else?)

Min_Retail refers to UNIT_PRC on the line or something else?  (If it is later discounted, does the system double check the spiff?  (doubt that..., should it?))

If none is found, does it then default to the ITM spiff?

Thanks,
Kent
----
**Further questions for Wayne**
A) If there is a spiff for itm and also for spiff calendar, which one should be recorded for the so_ln?
(Currently implemented as ITM)
B) If the answer is spiff calendar, then: if the min retail isn't met, would it then default to the itm's spiff?

Kent,
My answers follow:
A.      The spiff from the spiff calendar is recorded in SO_LN.
B.      If the minimum retail is not met it defaults to zero.
Thanks,
Wayne Liburd
----
What if there is another spiff amount recorded at the ITM level?

Kent,
On my test there was another spiff amount at the item level and it still entered zero.
----
I need to clarify/double check something:

If a product is added to an order a week after the order was written, would the SO.SO_WR_DT or the SO_LN.ADDON_WR_DT be used to look up the spiff calendar?

If a product is changed on an existing order, is today's date used?

Kent
----
Kent,
You should use the SO_LN.ADDON_WR_DT.
Yes, today’s date is used.
Thanks,
Wayne Liburd
----
Suppose line QTY is 2.  The spiff is still just copied straight over or is it multiplied by 2?

Kent,

The spiff is copied straight over.

Wayne

**Comment by kbower** [comment]
_**kbower** trac comment on Jul 1, 2010, 3:45:59 PM:_

_The big problem here is ability to process orders on a subsequent day._

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 24, 2011, 8:34:52 PM:_

Wayne,
If the spiff  END_DT  6/1/2011 and I sell one 6/1/2011 (same date as ending date), do I get the spiff?
Kent
----
Kent,
Yes.
Thanks,
Wayne Liburd

(Rob, please don't get upset because I point this out, but please try to remember that these effective dates that GERS uses all over have always been inclusive, so both ends of the query need an <= instead of just < (effectivedate <= date and date <= endingdate).  Thanks man!)

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 25, 2011, 9:16:27 AM:_

(In [[1120]](../commit/6f016ffc83e9c2247eb8537e09d50e93604b31ce))
 * fix problem where cannot save an order, caused by [r1118](../commit/e0bf171c12b72705f721f7df6335960513456b0d). references [#574](574)
 * fix "Truck stop system comment is being added at every order update".  fixes [#595](595).  See [#594](594)
 * removed some potentially destructive log statements.  See [#593](593)
 * Spiff changes (closes #218):
   * Spiff Calendar should take precedence over Product. fixes [#596](596)
   * Reference spiff calendar by addeddate of orderdetail or orderdate instead of today. (also [#596](596))
 * sql merge() may cause fk update that does not expire loaded related object.  
   * Fixed when not part of merge cascade
   * fixed [#597](597) (When changing products, spiff and commissiontype is copied from the old product.)
 * Order line edit screen: orderdetails, add addeddate

**Comment by kbower** [comment]
_**kbower** trac status change on Jan 25, 2011, 9:16:27 AM:_ **closed**
